src/network/networkd-ndisc.c

```diff
@@ -1029,6 +1029,56 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
         return r;
 }
 
+static int ndisc_process_zero_lifetime(Link *link, sd_ndisc_router *rt) {
+        NDiscCaptivePortal *cp;
+        bool updated = false;
+        NDiscDNSSL *dnssl;
+        NDiscRDNSS *rdnss;
+        Address *address;
+        Route *route;
+        int r, k;
+
+        assert(link);
+
+        SET_FOREACH(route, link->routes) {
+                if (route->source != NETWORK_CONFIG_SOURCE_NDISC)
+                        continue;
```

Comment:
Please also check if `route->provider` matches the router address.
There may be multiple routers in the network.